syntax = "proto3";
option go_package = "protos";

package main;

message Javascript {
  string name = 1;
  string version = 2;
  string code = 3;
}

message Struct {
    map<string, Value> fields = 1;
}

message ListValue {
    repeated Value values = 1;
}

enum ScopeEnum {
  HAND = 0;
  TEAM = 1;
  MODULE = 2;
  STINT = 3;
}

message Value {
  oneof kind {
    bool bool_value = 1;
    double number_value = 2;
    string string_value = 3;
    ListValue list_value = 4;
    Struct struct_value = 5;
  }
}

message Variable {
  int64 variable_definition_id = 1;
  Value value = 2;
}

message Hand {
  string name = 1;
}       

message Team {
  string name = 1;
  map<int32, Hand> members = 2;
}

message Stint {
  string name = 1;
  map<int32, Team> teams = 2;
}               

message Era {
  string name = 1;
}

message Stage {
  string name = 1;
}

message Context {
  map<string, Variable> variables = 1;
  Stint stint = 2;
  Era era = 3;
  Stage stage = 4;
}

message State {
  map<string, Variable> variables = 1;
  Era era = 2;
}

message JavascriptOp {
  Javascript script = 1;
  Context context = 2;
}

message Result {
  Value value = 1;
  State state = 2;
  double execution_time = 3;
}

service JavascriptEngine {
  rpc Run (JavascriptOp) returns (Result) {}
}
